blockchain/message_pool/src/msgpool.rs

```diff
@@ -776,11 +877,303 @@ where
+            warn!(
+                "encountered message from actor {} with nonce {} less than the current nonce {}",
+                actor,
+                m.sequence(),
+                cur_seq
+            );
+            skip += 1;
+            i += 1;
+            continue;
+        }
+        if m.sequence() != cur_seq {
+            break;
+        }
+        cur_seq += 1;
+        let min_gas = interpreter::price_list_by_epoch(ts.epoch())
+            .on_chain_message(m.marshal_cbor()?.len())
+            .total();
+        if m.gas_limit() < min_gas {
+            break;
+        }
+        gas_limit += m.gas_limit();
+
+        if gas_limit > types::BLOCK_GAS_LIMIT {
+            break;
+        }
+        let required = m.required_funds();
+        if balance < required {
+            break;
+        }
+        balance -= required;
+        let value = m.value();
+        if &balance >= value {
+            balance -= value;
+        }
+
+        let gas_reward = get_gas_reward(&m, base_fee);
+        rewards.push(gas_reward);
+        i += 1;
+    }
+    // check we have a sane set of messages to construct the chains
+    let msgs = if i > skip {
+        msgs[skip..i].to_vec()
+    } else {
+        return Ok(vec![]);
+    };
+
+    let new_chain = |m: SignedMessage, i: usize| -> MsgChain {
+        let gl = m.gas_limit();
+        let node = MsgChainNode {
+            msgs: vec![m],
+            gas_reward: rewards[i].clone(),
+            gas_limit: gl,
+            gas_perf: get_gas_perf(&rewards[i], gl),
+            eff_perf: 0.0,
+            bp: 0.0,
+            parent_offset: 0.0,
+            valid: true,
+            merged: false,
+        };
+        MsgChain {
+            index: 0,
+            chain: vec![node],
+        }
+    };
+
+    let mut chains = Vec::new();
+    let mut cur_chain = MsgChain::new();
+
+    for (i, m) in msgs.into_iter().enumerate() {
+        if i == 0 {
+            cur_chain = new_chain(m, i);
+            continue;
+        }
+        let gas_reward = cur_chain.curr().unwrap().gas_reward.clone() + &rewards[i];
+        let gas_limit = cur_chain.curr().unwrap().gas_limit + m.gas_limit();
+        let gas_perf = get_gas_perf(&gas_reward, gas_limit);
+
+        // try to add the message to the current chain -- if it decreases the gasPerf, then make a
+        // new chain
+        if gas_perf < cur_chain.curr().unwrap().gas_perf {
+            chains.push(cur_chain.clone());
+            cur_chain = new_chain(m, i);
+        } else {
+            let cur = cur_chain.curr_mut().unwrap();
+            cur.msgs.push(m);
+            cur.gas_reward = gas_reward;
+            cur.gas_limit = gas_limit;
+            cur.gas_perf = gas_perf;
+        }
+    }
+    chains.push(cur_chain);
+
+    // merge chains to maintain the invariant
+    loop {
+        let mut merged = 0;
+        for i in (1..chains.len()).rev() {
+            let (head, tail) = chains.split_at_mut(i);
```

Comment:
Won't this panic when indexing the 0th element of the tail when splitting a `chains.len()`, what am I missing here?

Reply:
This would, but we never split at chains.len(), its [1, chains.len() -1]

Reply:
ah, I had wrong idea of the split, didn't realize it was excluding the index